### PR TITLE
Implement str.capitalize() based on CPython

### PR DESCRIPTION
**Pull request by densmirn**

### CODE REVIEW

**Comment by densmirn** [comment]
Could you please add the PR to reviewing queue?

**Comment by stuartarchibald** [comment]
@densmirn thanks, queued.

File: numba/unicode.py [hunk 1]
```diff
@@ -1267,6 +1267,52 @@ def impl(a):
         return impl
 
 
+# https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Objects/unicodeobject.c#L10765-L10774    # noqa: E501
+@overload_method(types.UnicodeType, 'capitalize')
+def unicode_capitalize(data):
+    """Implements str.capitalize()"""
+    def impl(data):
+        # https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Objects/unicodeobject.c#L9737-L9759    # noqa: E501
```
Comment (stuartarchibald) [inline numba/unicode.py]: Impl seems good, though lacks an ASCII encoding fast-path, so performance will likely be poor in this case.

Comment (densmirn) [reply numba/unicode.py]: Added ascii capitalization.
@stuartarchibald could you please take the next round of review?

File: numba/unicode.py [hunk 2]
```diff
@@ -1267,6 +1267,62 @@ def impl(a):
...
+                    maxchars[0] = max(maxchar, m)
+                    _set_code_point(res, k, m)
+                    k += 1
+            return k
+
+        length = len(data)
+        if length == 0:
+            return _empty_string(data._kind, length, data._is_ascii)
+
+        if data._is_ascii:
```
Comment (stuartarchibald) [inline numba/unicode.py]: Missing CPython alg reference?

Comment (densmirn) [reply numba/unicode.py]: There is no such algorithm in CPython: https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Objects/unicodeobject.c#L10765-L10774. So there is no link here.

**Comment by densmirn** [comment]
@stuartarchibald could you please re-add the PR to the review queue?

File: numba/unicode.py (hunk 2 above)
Comment (stuartarchibald) [reply numba/unicode.py]: Is it not this https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Objects/bytes_methods.c#L361-L382 ?

Comment (densmirn) [reply numba/unicode.py]: According to the name that is Bytes method, not Unicode. I will insert the reference if you confirm  it.

File: numba/unicode.py [hunk 3]
```diff
@@ -1267,6 +1267,56 @@ def impl(a):
...
+    """Implements str.capitalize()"""
+    def impl(data):
+        length = len(data)
+        if length == 0:
+            return _empty_string(data._kind, length, data._is_ascii)
+
+        if data._is_ascii:
+            res = _empty_string(data._kind, length, 1)
+            code_point = _get_code_point(data, 0)
+            _set_code_point(res, 0, _Py_TOUPPER(code_point))
```
Comment (stuartarchibald) [inline numba/unicode.py]: In terms of efficiency, inside this section, it's probably faster to use the `_Py_IS<FOO>` methods to check if it's already `FOO` and then only translate the code point as needed? (like in the bytes method ref above).

File: numba/unicode.py (hunk 2 above)
Comment (stuartarchibald) [reply numba/unicode.py]: If you copied it/translated it from there then yes, please insert the reference.

Comment (densmirn) [reply numba/unicode.py]: I didn't copy/translated it from there. I did it by analogy with other implementations.

Comment (densmirn) [reply numba/unicode.py]: @stuartarchibald I will insert the reference, thanks.

**Pull request closed without merging**